File: transpiler/src/main/java/com/here/ivi/api/generator/cpp/TopologicalSort.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.generator.cpp;

import com.here.ivi.api.cli.TranspilerExecutionException;
import com.here.ivi.api.model.cpp.*;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class TopologicalSort {

  private final List<CppElement> elements;

  private final Set<String> fullyQualifiedNames;

  public TopologicalSort(List<CppElement> elements) {

    this.elements = elements;
    this.fullyQualifiedNames =
        elements.stream().map(e -> e.fullyQualifiedName).collect(Collectors.toSet());
  }

  /**
   * Do a topological sort based on Kahn's algorithm
   * https://en.wikipedia.org/wiki/Topological_sorting on the given structs and assign priorities to
   * structs so the most-basic structs are defined first to avoid compilation errors on C++.
   */
  public List<CppElement> sort() {

    List<CppElement> sortedElements = new LinkedList<>();

    // structs dependencies map
    Map<String, Set<String>> dependencies = buildDependencies();

    while (sortedElements.size() < elements.size()) {

      // find the first element with no dependency needed which is not in the sorted elements
      CppElement nextElement =
          elements
              .stream()
              .filter(
                  cppElement ->
                      dependencies.get(cppElement.fullyQualifiedName).isEmpty()
                          && !sortedElements.contains(cppElement))
              .findFirst()
              .orElse(null);

      if (nextElement == null) {
        throw new TranspilerExecutionException("Cycle detected in CPP elements dependencies.");
      }

      sortedElements.add(nextElement);

      // as dependency to "nextElement" is now fulfilled we must remove it from elements dependencies
      for (String name : fullyQualifiedNames) {
        dependencies.get(name).remove(nextElement.fullyQualifiedName);
      }
    }

    return sortedElements;
  }

  private Set<String> getTypeDependencies(CppTypeRef typeRef) {

    Set<String> dependencies = new HashSet<>();

    if (typeRef instanceof CppTemplateTypeRef) {
      CppTemplateTypeRef templateTypeRef = (CppTemplateTypeRef) typeRef;
      templateTypeRef.templateParameters.forEach(
          parameter -> dependencies.addAll(getTypeDependencies(parameter)));
    } else if (typeRef instanceof CppTypeDefRef) {
      CppTypeDefRef typeDefRef = (CppTypeDefRef) typeRef;
      dependencies.addAll(getTypeDependencies(typeDefRef.actualType));
    }

    if (fullyQualifiedNames.contains(typeRef.name)) {
      dependencies.add(typeRef.name);
    }

    return dependencies;
  }

  private Set<String> getElementDependencies(CppElement cppElement) {

    if (cppElement instanceof CppTypedElement) {
      CppTypedElement cppTypedElement = (CppTypedElement) cppElement;
      Set<String> dependencies = getTypeDependencies(cppTypedElement.type);
      if (fullyQualifiedNames.contains(cppTypedElement.type.name)) {
        dependencies.add(cppTypedElement.type.name);
      }
      return dependencies;
    }

    if (cppElement instanceof CppStruct) {
      CppStruct cppStruct = (CppStruct) cppElement;
      return getStructDependencies(cppStruct);
    }

    if (cppElement instanceof CppUsing) {
      CppUsing cppUsing = (CppUsing) cppElement;
      return getTypeDependencies(cppUsing.definition);
    }

    return Collections.emptySet();
  }

  private Set<String> getStructDependencies(CppStruct cppStruct) {

    return Stream.concat(
            cppStruct
                .inheritances
                .stream()
                .map(cppInheritance -> cppInheritance.parent.fullyQualifiedName)
                .filter(fullyQualifiedNames::contains),
            cppStruct
                .fields
                .stream()
                .flatMap(cppField -> getElementDependencies(cppField).stream()))
        .collect(Collectors.toSet());
  }

  private Map<String, Set<String>> buildDependencies() {

    return elements
        .stream()
        .collect(
            Collectors.toMap(element -> element.fullyQualifiedName, this::getElementDependencies));
  }
}

File: gluecodium/src/main/java/com/here/gluecodium/generator/cpp/TopologicalSort.kt
/*
 * Copyright (C) 2016-2020 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium.generator.cpp

import com.here.gluecodium.cli.GluecodiumExecutionException
import com.here.gluecodium.model.lime.LimeFunction
import com.here.gluecodium.model.lime.LimeGenericType
import com.here.gluecodium.model.lime.LimeLambda
import com.here.gluecodium.model.lime.LimeNamedElement
import com.here.gluecodium.model.lime.LimeStruct
import com.here.gluecodium.model.lime.LimeTypeAlias
import com.here.gluecodium.model.lime.LimeTypeRef
import com.here.gluecodium.model.lime.LimeTypedElement

internal class TopologicalSort(private val elements: List<LimeNamedElement>) {

    private val fullNames = elements.map { it.fullName }.toSet()

    /**
     * Do a topological sort based on Kahn's algorithm
     * https://en.wikipedia.org/wiki/Topological_sorting on the given structs and assign priorities
     * to structs so the most-basic structs are defined first to avoid compilation errors on C++.
     */
    fun sort(): List<LimeNamedElement> {
        val dependencies = elements.associateBy(
            { it.fullName },
            { getElementDependencies(it).toMutableSet() }
        )

        val sortedElements = mutableListOf<LimeNamedElement>()
        while (sortedElements.size < elements.size) {
            // Find the first element with no dependency needed which is not in the sorted elements.
            val nextElement = elements.firstOrNull {
                val dependencySet = dependencies[it.fullName]
                dependencySet != null && dependencySet.isEmpty() && !sortedElements.contains(it)
            } ?: throw GluecodiumExecutionException("Cycle detected in C++ elements dependencies.")

            sortedElements.add(nextElement)

            // As dependency to "nextElement" is now fulfilled we must remove it from elements
            // dependencies.
            for (name in fullNames) {
                dependencies[name]?.remove(nextElement.fullName)
            }
        }

        return sortedElements
    }

    private fun getTypeRefDependencies(limeTypeRef: LimeTypeRef): List<String> =
        when (val limeType = limeTypeRef.type) {
            is LimeGenericType -> limeType.childTypes.flatMap { getTypeRefDependencies(it) }
            else -> emptyList()
        } + listOfNotNull(limeTypeRef.type.fullName.takeIf { fullNames.contains(it) })

    private fun getElementDependencies(limeElement: LimeNamedElement): Set<String> =
        when (limeElement) {
            is LimeTypedElement -> getTypeRefDependencies(limeElement.typeRef)
            is LimeStruct -> (limeElement.fields + limeElement.functions + limeElement.constants)
                .flatMap { getElementDependencies(it) }
            is LimeTypeAlias -> getTypeRefDependencies(limeElement.typeRef)
            is LimeFunction -> (
                limeElement.parameters.map { it.typeRef } +
                    listOfNotNull(limeElement.returnType.typeRef, limeElement.exception?.errorType)
                )
                .flatMap { getTypeRefDependencies(it) }
            is LimeLambda -> (limeElement.parameters.map { it.typeRef } + limeElement.returnType.typeRef)
                .flatMap { getTypeRefDependencies(it) }
            else -> emptyList()
        }.toSet() - limeElement.fullName
}
